roc.go

```diff
@@ -43,35 +44,42 @@ func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []floa
 	if !sort.Float64sAreSorted(y) {
 		panic("stat: input must be sorted")
 	}
-
-	var incWidth, tol float64
-	if n == 0 {
-		if len(y) == 0 {
-			return nil, nil
-		}
-		tpr = make([]float64, len(y)+1)
-		fpr = make([]float64, len(y)+1)
-	} else {
-		if n < 2 {
-			panic("stat: cannot calculate fewer than 2 points on a ROC curve")
-		}
-		if len(y) == 0 {
-			return nil, nil
-		}
-		tpr = make([]float64, n)
-		fpr = make([]float64, n)
-		incWidth = (y[len(y)-1] - y[0]) / float64(n-1)
-		tol = y[0] + incWidth
-		if incWidth == 0 {
-			tpr[n-1] = 1
-			fpr[n-1] = 1
-			return
+	if !sort.Float64sAreSorted(cutoffs) {
+		panic("stat: input must be sorted")
+	}
+	if len(y) == 0 {
+		return nil, nil
+	}
+	var bin int
+	if len(cutoffs) == 0 {
```

Comment:
cutoffs can be longer than len(y)+1 though -- this will simply result in repeated values on the ROC curve. What this should be checking for is if cutoffs is nil or empty, which is what len(cutoffs)==0 achieves -- given that this is the case though you are right I should check the cap in case calling make can be avoided.